File: leet_LIST_0138_copy_list.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    E := make( map[ *Node ] *Node )
    curr := head
    for curr != nil {
        temp := & Node { Val: curr.Val}
        E[curr] = temp
        curr = curr.Next
    }
    curr = head
    for curr != nil {
        ptr := E[curr]
        ptr.Next = E[curr.Next]
        ptr.Random = E[curr.Random]
        curr = curr.Next
    }
    return E[head]
}

File: leet_LIST_0138_copy_list.ts
/**
 * Definition for Node.
 * class Node {
 *     val: number
 *     next: Node | null
 *     random: Node | null
 *     constructor(val?: number, next?: Node, random?: Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */

function copyRandomList(head: Node | null): Node | null {
    const E: Map<Node | null, Node | null> = new Map();

    let curr: Node | null = head;

    while (curr !== null) {
        const temp: Node | null = new Node(curr.val);
        E.set(curr, temp);
        curr = curr.next;
    }

    curr = head;

    while (curr !== null) {
        const ptr: Node | null = E.get(curr);
        if (ptr !== undefined) {
            ptr.next = E.get(curr.next) || null;
            ptr.random = E.get(curr.random) || null;
        }
        curr = curr.next;
    }

    return E.get(head) || null;
};
